state/application.go

```diff
@@ -1180,6 +1182,40 @@ func (a *Application) SetCharm(cfg SetCharmConfig) (err error) {
 	// we don't need to check that this is a charm.LXDProfiler, as we can
 	// state that the function exists.
 	if profile := cfg.Charm.LXDProfile(); profile != nil {
+		// Check if the controller agent tools version is greater than the
+		// version we support for the new LXD profiles.
+		// Then check all the units, to see what their agent tools versions is
+		// so that we can ensure that everyone is aligned. If the units version
+		// is to low (i.e. less than the 2.5.2 epoch), then show an error
+		// message that the operator should upgrade to receive the latest
+		// LXD Profile changes.
+		if featureflag.Enabled(feature.InstanceMutater) {
+			epoch := version.Number{Major: 2, Minor: 5, Patch: 2}
+			tools, err := a.AgentTools()
+			if err != nil {
+				return errors.Annotate(err, "cannot retrieve agent tools")
+			}
+			if ver := tools.Version; ver.Compare(epoch) > 0 {
+				units, err := a.AllUnits()
+				if err != nil {
+					return errors.Annotate(err, "cannot retrieve units")
+				}
+				for _, unit := range units {
+					unitTools, err := unit.AgentTools()
+					if err != nil {
+						return errors.Annotate(err, "cannot retrieve unit agent tools")
+					}
+					if unitVer := unitTools.Version; unitVer.Compare(epoch) <= 0 {
+						return errors.Errorf(
+							"Unable to upgrade LXDProfile charms with the current version. "+
```

Comment:
Let's be more specific about what needs to be upgraded, perhaps "Unable to upgrade LXDProfile charms with the current model version.  Please run juju upgrade-juju to upgrade the current model to match your controller."

Reply:
Yeah, I was really struggling with the correct wording here.